### PR TITLE
[SPARK-24952][SQL] Support LZMA2 compression by Avro datasource

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

In the PR, I propose to support `LZMA2` (`XZ`) and `BZIP2` compressions by `AVRO` datasource  in write since the codecs may have better characteristics like compression ratio and speed comparing to already supported `snappy` and `deflate` codecs.

## How was this patch tested?

It was tested manually and by an existing test which was extended to check the `xz` and `bzip2` compressions.

### CODE REVIEW

**Comment by MaxGekk** [comment]
jenkins, retest this, please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1449,6 +1451,16 @@ object SQLConf {
     .intConf
     .checkValues((1 to 9).toSet + Deflater.DEFAULT_COMPRESSION)
     .createWithDefault(Deflater.DEFAULT_COMPRESSION)
+
+  val AVRO_XZ_LEVEL = buildConf("spark.sql.avro.xz.level")
+    .doc("Compression level for the XZ codec used in writing of AVRO files. " +
+      "Valid value must be in the range of from 0 to 9 inclusive: " +
+      "0-3 for fast with medium compression, 4-6 are fairly slow levels with high compression. " +
+      "The levels 7-9 are like the level 6 but use bigger dictionaries and have higher " +
+      "compressor and decompressor memory requirements. Default level is 6.")
```
Comment (felixcheung) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: use `LZMA2Options.PRESET_DEFAULT` in place of literal `6`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -1449,6 +1451,16 @@ object SQLConf {
     .intConf
     .checkValues((1 to 9).toSet + Deflater.DEFAULT_COMPRESSION)
     .createWithDefault(Deflater.DEFAULT_COMPRESSION)
+
+  val AVRO_XZ_LEVEL = buildConf("spark.sql.avro.xz.level")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Just for clarification, @MaxGekk, this configuration is not for keeping the configuration in spark-avro (as a third party) but something you here newly propose?

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Yes, it is a new feature. The third party doesn't support `LZMA2` codec. The last one has much better compression than `snappy`/`deflate` even on level `0-3`. The default level `6` covers most of use cases actually when users need to store data in AVRO format permanently as an archive. If you strongly against of the config, I will remove it.

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: > If you faced a case that this configuration is useful, im okay but if not, I would suggest to add it later when it's needed.

The config is not so useful as `LZMA2` codec itself. Its default value covers the use case. I will remove the config from the PR.

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 3]
```diff
@@ -117,28 +117,19 @@ private[avro] class AvroFileFormat extends FileFormat with DataSourceRegister {
...
+      case DEFLATE_CODEC =>
         val deflateLevel = spark.sessionState.conf.avroDeflateLevel
-        log.info(s"compressing Avro output using deflate (level=$deflateLevel)")
-        job.getConfiguration.setBoolean(COMPRESS_KEY, true)
-        job.getConfiguration.set(AvroJob.CONF_OUTPUT_CODEC, DataFileConstants.DEFLATE_CODEC)
         job.getConfiguration.setInt(AvroOutputFormat.DEFLATE_LEVEL_KEY, deflateLevel)
-
-      case unknown: String =>
-        log.error(s"unsupported compression codec $unknown")
+        (DEFLATE_CODEC, s" (level = $deflateLevel)")
```
Comment (HyukjinKwon) [inline external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: @MaxGekk, how about just log another info here and remove `levelInfo` return here? for instance,

```
log.info("Avro compression level 6 will be used for deflate codec.")
```

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

File: external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala (hunk 3 above)
Comment (MaxGekk) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Conversion of `-1` to `6` is performed in somewhere inside of `zlib`. Do you want to hard code `6` in log message?

Comment (HyukjinKwon) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: Oh, I mean I thought it's a bit odd this case match returns levelInfo abd wondering if we can get rid of it.

Comment (MaxGekk) [reply external/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala]: I have refactored this piece of code.

**Comment by MaxGekk** [comment]
jenkins, retest this, please

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**